core/src/main/java/org/elasticsearch/action/search/SearchExecutionStatsCollector.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.search;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.node.ResponseCollectorService;
+import org.elasticsearch.search.SearchPhaseResult;
+import org.elasticsearch.search.query.QuerySearchResult;
+
+import java.util.Objects;
+
+/**
+ * A wrapper of search action listeners (search results) that unwraps the query
+ * result to get the piggybacked queue size and service time EWMA, adding those
+ * values to the coordinating nodes' {@code ResponseCollectorService}.
+ */
+public class SearchExecutionStatsCollector implements ActionListener<SearchPhaseResult> {
+
+    private final SearchActionListener<SearchPhaseResult> listener;
+    private final ResponseCollectorService collector;
+    private final long startNanos;
+
+    public SearchExecutionStatsCollector(SearchActionListener<SearchPhaseResult> listener,
```

Comment:
please instead of exposing the searchShardTarget from `SearchActionListener` can you add a 3rd arg that is the nodeID. we can get it from the passed in connection we talk to instead. then `SearchActionListener` can also be a `ActionListener` instead